pulp_container/app/viewsets.py

```diff
@@ -435,3 +457,403 @@ class ContainerDistributionViewSet(BaseDistributionViewSet):
+            size = artifact.size
+        else:
+            size = 0
+        headers = {'Docker-Distribution-Api-Version': 'registry/2.0',
+                   'Docker-Content-Digest': manifest.digest,
+                   'Location': '/v2/{path}/manifests/{digest}'.format(path=path,
+                                                                      digest=manifest.digest),
+                   'Content-Length': size
+                   }
+        super().__init__(headers=headers, status=status)
+
+
+class BlobResponse(Response):
+    """
+    An HTTP response class for returning Blobs.
+    """
+
+    def __init__(self, blob, path, status, request, send_body=False):
+        """
+        Args:
+            blob (pulp_container.app.models.Blob): A Blob model used to generate the response.
+            path (str): The base_path of the ContainerDistribution (Container repository name)
+            request (rest_framework.request.Request): Request object not used by this
+                implementation of Response.
+            status (int): Status code to send with the response.
+            send_body (bool): Whether a body should be sent with the response or just the
+                headers.
+        """
+        artifact = blob._artifacts.get()
+        size = artifact.size
+
+        log.info('digest: {digest}'.format(digest=blob.digest))
+        headers = {'Docker-Distribution-Api-Version': 'registry/2.0',
+                   'Docker-Content-Digest': blob.digest,
+                   'Location': '/v2/{path}/blobs/{digest}'.format(path=path, digest=blob.digest),
+                   'Etag': blob.digest,
+                   'Range': '0-{offset}'.format(offset=int(size)),
+                   'Content-Length': size,
+                   'Content-Type': 'application/octet-stream',
+                   'Connection': 'close'
+                   }
+        super().__init__(headers=headers, status=status)
+
+
+class VersionView(APIView):
+    """
+    Handles requests to the /v2/ endpoint.
+    """
+
+    # allow anyone to access
+    authentication_classes = []
+    permission_classes = []
+
+    def get(self, request):
+        """Handles GET requests for the /v2/ endpoint."""
+        headers = {'Docker-Distribution-Api-Version': 'registry/2.0',
+                   }
+        return Response(data={}, headers=headers)
+
+
+class BlobUploads(ViewSet):
+    """
+    The ViewSet for handling uploading of blobs.
+    """
+
+    model = models.Upload
+    queryset = models.Upload.objects.all()
+
+    # allow anyone to access
+    authentication_classes = []
+    permission_classes = []
+
+    content_range_pattern = re.compile(r'^(?P<start>\d+)-(?P<end>\d+)$')
+
+    def create(self, request, path):
+        """
+        This methods handles the creation of an upload.
+        """
+        distribution = get_object_or_404(models.ContainerDistribution, base_path=path)
+        if distribution.repository:
+            repository = distribution.repository
```

Comment:
NB when creating a distribution we can point only a repo to it and NOT repo_version. This leads to the questions whether we separate somehow a 'mirror' repo and a 'puhs' repo